test/server/drain_manager_impl_test.cc

```diff
@@ -151,8 +159,398 @@ TEST_P(DrainManagerImplTest, DrainDeadlineProbability) {
+        EXPECT_LT(delay.count(), 1001);
+
+        // Validate that our wait times are spread out (within some small error)
+        EXPECT_THAT(delay.count(), AllOf(Ge(i * step - 1), Le(i * step + 1)));
+      }));
+
+      cb_handles[i] = drain_manager.addOnDrainCloseCb(cb.AsStdFunction());
+    }
+    drain_manager.startDrainSequence([] {});
+  }
+
+  EXPECT_TRUE(drain_manager.draining());
+}
+
+// Test gradual draining when the number of callbacks does not evenly divide into
+// the drain time.
+TEST_F(DrainManagerImplTest, OnDrainCallbacksNonEvenlyDividedSteps) {
+  constexpr int num_cbs = 30;
+  ON_CALL(server_.options_, drainStrategy()).WillByDefault(Return(Server::DrainStrategy::Gradual));
+  ON_CALL(server_.options_, drainTime()).WillByDefault(Return(std::chrono::seconds(1)));
+
+  DrainManagerImpl drain_manager(server_, envoy::config::listener::v3::Listener::DEFAULT,
+                                 server_.dispatcher());
+
+  {
+    // Register callbacks (store in array to keep in scope for test)
+    std::array<testing::MockFunction<void(std::chrono::milliseconds)>, num_cbs> cbs;
+    std::array<Common::CallbackHandlePtr, num_cbs> cb_handles;
+    for (auto i = 0; i < num_cbs; i++) {
+      auto& cb = cbs[i];
+      auto step = 250.0 / num_cbs;
+      EXPECT_CALL(cb, Call(_)).WillRepeatedly(Invoke([i, step](std::chrono::milliseconds delay) {
+        // Everything should happen within the first 1/4 of the drain time
+        EXPECT_LT(delay.count(), 251);
+
+        // Validate that our wait times are spread out (within some small error)
+        EXPECT_THAT(delay.count(), AllOf(Ge(i * step - 1), Le(i * step + 1)));
+      }));
+
+      cb_handles[i] = drain_manager.addOnDrainCloseCb(cb.AsStdFunction());
+    }
+
+    drain_manager.startDrainSequence([] {});
+  }
+
+  EXPECT_TRUE(drain_manager.draining());
+}
+
+// Validate the expected behavior when a drain-close callback is registered after draining has begun
+// with a Gradual drain strategy (should be called with delay between 0 and maximum)
+TEST_F(DrainManagerImplTest, RegisterCallbackAfterDrainBeginGradualStrategy) {
+  ON_CALL(server_.options_, drainStrategy()).WillByDefault(Return(Server::DrainStrategy::Gradual));
+  ON_CALL(server_.options_, drainTime()).WillByDefault(Return(std::chrono::seconds(1)));
+
+  DrainManagerImpl drain_manager(server_, envoy::config::listener::v3::Listener::DEFAULT,
+                                 server_.dispatcher());
+
+  testing::MockFunction<void(std::chrono::milliseconds)> cb_before_drain;
+  testing::MockFunction<void(std::chrono::milliseconds)> cb_after_drain;
+
+  EXPECT_CALL(cb_before_drain, Call(_));
+  EXPECT_CALL(cb_after_drain, Call(_)).WillOnce(Invoke([](std::chrono::milliseconds delay) {
+    // Validate that callbacks after the drain sequence has started are called with a random value
+    // between 0 (immediate) and the max drain window.
+    EXPECT_THAT(delay.count(), Ge(0));
+    EXPECT_THAT(delay.count(), Le(1000));
```

Comment:
The actual behavior is that this should be less than the remaining time, right? Is it worth advancing the time after starting the drain to validate that this upper bound is limited in that way?

Reply:
Ah, good point! I changed the test for the callbacks registered after drain start to first advance 10ms and then 1000ms. And adjusting the assertions for expected time-range to match (0-990, and exactly 0).